src/Mod/Sketcher/App/planegcs/Constraints.cpp

```diff
@@ -2494,6 +2494,212 @@ double ConstraintAngleViaPoint::grad(double *param)
     deriv += ( (-n2.dx) * n2.y / pow(n2.length(), 2)  +  n2.dy * n2.x / pow(n2.length(), 2) );
 
 
+// use numeric for testing
+#if 0
+    double const eps = 0.00001;
+    double oldparam = *param;
+    double v0 = this->error();
+    *param += eps;
+    double vr = this->error();
+    *param = oldparam - eps;
+    double vl = this->error();
+    *param = oldparam;
+    //If not nasty, real derivative should be between left one and right one
+    double numretl = (v0-vl)/eps;
+    double numretr = (vr-v0)/eps;
+    assert(deriv <= std::max(numretl,numretr) );
+    assert(deriv >= std::min(numretl,numretr) );
+#endif
+
+    return scale * deriv;
+}
+
+// --------------------------------------------------------
+// ConstraintAngleViaPointAndParam
+ConstraintAngleViaPointAndParam::ConstraintAngleViaPointAndParam(Curve& acrv1, Curve& acrv2, Point p, double* cparam, double* angle)
+{
+    pvec.push_back(angle);
+    pvec.push_back(p.x);
+    pvec.push_back(p.y);
+    pvec.push_back(cparam);
+    acrv1.PushOwnParams(pvec);
+    acrv2.PushOwnParams(pvec);
+    crv1 = acrv1.Copy();
+    crv2 = acrv2.Copy();
+    origpvec = pvec;
+    pvecChangedFlag = true;
+    rescale();
+}
+
+ConstraintAngleViaPointAndParam::~ConstraintAngleViaPointAndParam()
+{
+    delete crv1; crv1 = nullptr;
+    delete crv2; crv2 = nullptr;
+}
+
+void ConstraintAngleViaPointAndParam::ReconstructGeomPointers()
+{
+    int cnt=0;
+    cnt++;//skip angle - we have an inline function for that
+    poa.x = pvec[cnt]; cnt++;
+    poa.y = pvec[cnt]; cnt++;
+    cnt++;//skip cparam
+    crv1->ReconstructOnNewPvec(pvec,cnt);
+    crv2->ReconstructOnNewPvec(pvec,cnt);
+    pvecChangedFlag=false;
+}
+
+ConstraintType ConstraintAngleViaPointAndParam::getTypeId()
+{
+    return AngleViaPointAndParam;
+}
+
+void ConstraintAngleViaPointAndParam::rescale(double coef)
+{
+    scale = coef * 1.;
+}
+
+double ConstraintAngleViaPointAndParam::error()
+{
+    if (pvecChangedFlag)
+        ReconstructGeomPointers();
+    double ang = *angle();
+    DeriVector2 n1 = crv1->CalculateNormal(cparam());
+    DeriVector2 n2 = crv2->CalculateNormal(poa);
+
+    // rotate n1 by angle
+    DeriVector2 n1r(n1.x * cos(ang) - n1.y * sin(ang), n1.x * sin(ang) + n1.y * cos(ang));
```

Comment:
Do you think it makes sense to extend DeriVector2 to rotate by an angle, instead of hardcoding the rotation and derivative?

I am not sure exactly what is intended here, so I may be wrong. Please take a look at the sign of the derivative. If what is intended here is that the second argument is the partial of the first with respect to the angle, it appears the sign is reversed (dcos=-sin, dsin=cos).

Reply:
No comments (yet) on the first paragraph. Derivector math needs some thinking. I recently happened to come across a video about "dual numbers" which they called "cousins" of complez numbers that have use in these" value + derivative" calculations.

In the line, the arguments are x and y components of the rotated vector. I believe derivative information is lost when creating n1r (take a look at the constructor).

Reply:
My bad. I know derivector constructor takes coordinates and then derivatives, but somehow I did not realise that it takes x and y and then their derivatives. The fact that sin and cos derivatives are cos and -sin did not help me realise my mistake. Sorry for that.

I remember DeepSOIC was into "dual numbers" when he created DeriVector2.

I am also sorry I did not manage to progress on this review. I will try to carry on tomorrow.